File: Buoi_2/DoBubble.cpp
#include <iostream>
#include <vector>
#include <chrono> // Thư viện để đo thời gian

using namespace std;
using namespace chrono; // Để dễ gọi các đối tượng từ chrono

// Hàm Bubble Sort
void bubbleSort(vector<int>& arr) {
    int n = arr.size();
    bool swapped;
    for (int i = 0; i < n - 1; i++) {
        swapped = false;
        for (int j = 0; j < n - 1 - i; j++) { // Giảm dần phạm vi sắp xếp
            if (arr[j] > arr[j + 1]) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                swapped = true;
            }
        }
        if (!swapped)
            break;
    }
}

int main() {
    vector<int> arr = {5, 2, 9, 1, 5};

    // Đo thời gian bắt đầu
    auto start = high_resolution_clock::now();

    bubbleSort(arr);

    // Đo thời gian kết thúc
    auto end = high_resolution_clock::now();

    // Tính toán thời gian thực thi
    auto duration = duration_cast<microseconds>(end - start);
    cout << "Thoi gian thuc hien Bubble Sort: " << duration.count() << " microseconds" << endl;

    // In kết quả sau khi sắp xếp
    cout << "Ket qua sau Bubble Sort: ";
    for (int x : arr) 
        cout << x << " ";
    cout << endl;

    return 0;
}

File: Buoi_2/DoBubble.py
import time

def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1): 
        swapped = False
        for j in range(n - 1 - i):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j] 
                swapped = True
        if not swapped:
            break

if __name__ == "__main__": 
    arr = [5, 2, 9, 1, 5]

    # Ghi lại thời gian bắt đầu
    start_time = time.time()
    
    bubble_sort(arr)
    
    # Ghi lại thời gian kết thúc
    end_time = time.time()
    
    # Tính toán và in ra thời gian thực hiện
    execution_time = end_time - start_time
    print("Kết quả sau Bubble Sort:", arr)
    print(f"Thời gian thực hiện: {execution_time} giây")
